src/compiler/checker.ts

```diff
@@ -34523,6 +34523,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
             return getRegularTypeOfLiteralType(exprType);
         }
+        const links = getNodeLinks(node);

```

Comment:
```suggestion
function checkAssertionWorker(node: JSDocTypeAssertion | AssertionExpression, checkMode: CheckMode | undefined) {
        if (isConstTypeReference(type)) {
            if (!isValidConstAssertionArgument(expression)) {
                error(expression, Diagnostics.A_const_assertions_can_only_be_applied_to_references_to_enum_members_or_string_number_boolean_array_or_object_literals);
            }
            return getRegularTypeOfLiteralType(checkExpression(expression, checkMode));
        }
        checkSourceElement(type);
        checkNodeDeferred(node);
        return getTypeFromTypeNode(type);
 }
```
and no changes in `checkAssertionDeferred` from current `main` should result in the same number of `checkExpression` calls as the unsafe cache here, but without needing an unsafe cache, is what I'm trying to say. :P

Reply:
So getting rid of `assertionExpressionType` on `NodeLinks` entirely, just only doing `checkExpression` in `const` contexts and in the deferred check. Makes sense.